### ISSUE TITLE
Fetching tags from docker registry is limited to 20 pages

### ISSUE THREAD

**Issue by eyalzek** [renovatebot/renovate#9176]
**What would you like Renovate to be able to do?**

At the moment pagination is limited to 20 pages:
https://github.com/renovatebot/renovate/blob/master/lib/datasource/docker/index.ts#L458-L464
this proves problematic in certain registries/repositories as this means that renovate might not fetch _all_ tags, but only a subset. Having the option to remove/set the value for this limit will help mitigating this problem.

**Did you already have any implementation ideas?**

As per the discussion in https://github.com/renovatebot/renovate/discussions/9159, possible solutions would be to add a configuration option to change the default page limit/maximum tags limit.

I can attempt to write it myself .

**Comment by rarkins** [comment]
Unlimited or user-controlled pagination limits have the following risks:
- Overloading ourselves (could be thousands of pages)
- Overloading the remote resource (they may have no intention of anyone ever fetching *all* tags)
- Getting blocked by the remote resource
- A very easy attack vector

Therefore controlling pagination limits should be an admin/self-hosted configuration option only.

**Comment by eyalzek** [comment]
> Unlimited or user-controlled pagination limits have the following risks:
> 
>     * Overloading ourselves (could be thousands of pages)
> 
>     * Overloading the remote resource (they may have no intention of anyone ever fetching _all_ tags)
> 
>     * Getting blocked by the remote resource
> 
>     * A very easy attack vector
> 
> 
> Therefore controlling pagination limits should be an admin/self-hosted configuration option only.

Makes sense. Will bumping the 20 page default be an option? As the tag limit is otherwise set to 10000 (https://github.com/renovatebot/renovate/blob/master/lib/datasource/docker/index.ts#L450) it might be fine to set the page limit at around 50?

**Comment by rarkins** [comment]
The limit is not so much because we are worried about the total items returned but because of total requests. Paginating 50 or more requests - potentially per image - is really not good. We ideally wouldn't want to paginate more without a better way of caching. e.g. if quay.io could return items in the order of most recent then it wouldn't be such a problem, because we could stop fetching as soon as we start matching up with our cached results (likely on the first page).

**Comment by viceice** [comment]
Opened an issue for quay.io to fix the ordering

https://issues.redhat.com/browse/PROJQUAY-1741

**Comment by viceice** [comment]
Ok, made some more tests, docker hub returns tags alpha sorted, so we can get the issue there too, when there are too many tags.

**Comment by eyalzek** [comment]
It looks like this is not an issue with the v1 api:
`curl -sL  'https://quay.io/api/v1/repository/bitnami/sealed-secrets-controller/tag/' | jq`

this gets the latest tags as opposed to:
`curl -sL  'https://quay.io/v2/bitnami/sealed-secrets-controller/tags/list'  | jq`

**Comment by jwnmulder** [comment]
I'm having the same problem for another image on quay.io. Latest version is 3.2.1 but not found.

https://quay.io/repository/fairwinds/polaris?tab=tags (~1300 tags)

`getLabels(https://quay.io, fairwinds/polaris, cc6b64ded012d02a33bf5d520417f67ed484dc67)`

**Comment by rarkins** [comment]
Perhaps we can switch to the v1 api for tag list retrieval for quay.io. This may be preferable to paginating unlimited pages until we reach the end.

**Comment by eyalzek** [comment]
> Perhaps we can switch to the v1 api for tag list retrieval for quay.io. This may be preferable to paginating unlimited pages until we reach the end.

That's what the quay ui uses, so it might be a valid option.

**Comment by rarkins** [comment]
Happy to accept a PR for this approach if anyone has the time. Ie if the registry is quay.up then fetch tags using the v1 API. Let's keep to 20 pages and see how it goes. 

I'm not 100% sure if we can keep remaining functions to use the existing API calls though. May need testing.

**Comment by eyalzek** [comment]
I started a draft PR https://github.com/renovatebot/renovate/pull/9362 - untested at the moment. I guess extending the tests would be necessary, please let me know if I missed anything obvious.

**Comment by giokara-oqton** [comment]
Interested if there is any update on this? Seems like the PR is stale?

**Comment by eyalzek** [comment]
> Interested if there is any update on this? Seems like the PR is stale?

I don't have much time to focus on this at the moment. Feel free to pick it up.

**Comment by giokara-oqton** [comment]
The fix is ready (tested in https://github.com/oqton/test-renovate-docker2/pull/4). Can I get access to the repo in order to commit the necessary changes?

**Comment by viceice** [comment]
You can simply open a new pr and add a comment which other pr it will replace
